server/src/main/java/org/elasticsearch/reservedstate/service/ReservedStateUpdateTask.java

```diff
@@ -100,6 +101,36 @@ protected ClusterState execute(final ClusterState currentState) {
             }
         }
 
+        checkAndThrowOnError(errors, currentState, reservedStateVersion);
+
+        // Run the postTransform loop for any handlers that need to perform updates to other things than the cluster state
+        for (var handlerName : orderedHandlers) {
+            ReservedClusterStateHandler<?> handler = handlers.get(handlerName);
+            try {
+                // We first try to fetch the existing keys from what we have in the builder if transform saved any keys
+                Set<String> existingKeys = keysForHandler(reservedMetadataBuilder.getHandler(handlerName), existingMetadata, handlerName);
+                // postTransform doesn't modify the cluster state
+                Set<String> newKeys = handler.postTransform(reservedState.get(handlerName), state, existingKeys);
+                reservedMetadataBuilder.putHandler(new ReservedStateHandlerMetadata(handlerName, newKeys));
+            } catch (Exception e) {
+                errors.add(format("Error processing %s state change: %s", handler.name(), stackTrace(e)));
+            }
+        }
+
+        // Most error should be caught in the transform phase, however we check if the post transform encountered any
+        checkAndThrowOnError(errors, currentState, reservedStateVersion);
+
+        // remove the last error if we had previously encountered any
+        reservedMetadataBuilder.errorMetadata(null);
```

Comment:
I understand it is existing code. But I am curious about what other errors we could have here since we already check and throw previously?

Reply:
So when we fail some file processing, we report the errors in the logs, however I also have a cluster state update task that stores metadata into the cluster state indicating the errors encountered by the most recent processing.

Having this error metadata helps with building external support for verifying if the supplied settings.json file can be applied. Namely, instead of building pipelines to extract log information and look for specific text patters, we can build REST API that extracts the latest file processing information from the cluster state. I don't have this API yet, but it's in plan to be built.

So when we successfully process a file, we nullify the last error state from prior processing, meaning, the latest file processed (with certain version) was saved correctly.

For example:

```
file settings version: 22 -> saved OK, no error metadata, last applied settings version is 22
file settings version: 23 -> error while processing file, last applied settings version is 22, error version: 23, errors:....
file settings version: 24 -> saved OK, cleanup error metadata, last applied settings version is 24, error: null
```